src/curl_jolokia.c

```diff
@@ -0,0 +1,1100 @@
+ *
+ * Authors:
+ *   Doug MacEachern <dougm at hyperic.com>
+ *   Florian octo Forster <octo at collectd.org>
+ *   Wilfried dothebart Goesgens <dothebart at citadel.org>
+ **/
+
+#include "common.h"
+#include "configfile.h"
+#include "plugin.h"
+#include "utils_complain.h"
+#include "utils_llist.h"
+#include "collectd.h"
+
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+
+#include <curl/curl.h>
+
+#include <yajl/yajl_parse.h>
+#if HAVE_YAJL_YAJL_VERSION_H
+#include <yajl/yajl_version.h>
+#endif
+
+#if defined(YAJL_MAJOR) && (YAJL_MAJOR > 1)
+#define HAVE_YAJL_V2 1
+#endif
+
+#define CJO_DEFAULT_HOST "localhost"
+
+#if HAVE_YAJL_V2
+typedef size_t yajl_len_t;
+#else
+typedef unsigned int yajl_len_t;
+#endif
+
+struct cjo_attribute_s /* {{{ */
+{
+  char *attribute_name;
+  char *attribute_match;
+  size_t attribute_match_len;
+  char *type;
+};
+typedef struct cjo_attribute_s cjo_attribute_t;
+/* }}} */
+
+struct cjo_bean_s /* {{{ */
+{
+  char *bean_name;
+  char *mbean_match;
+  char *mbean_namespace;
+  size_t mbean_match_len;
+  llist_t *attributes; /* list of cjo_attribute_t */
+};
+typedef struct cjo_bean_s cjo_bean_t;
+/* }}} */
+
+typedef enum _expect_token { eNone = 0, eValue, eMBean } cjo_expect_token;
+
+/* {{{ */
+struct cjo_attribute_values_s {
+  const char *json_value;
+  yajl_len_t json_value_len;
+  const char *json_name;
+  yajl_len_t json_name_len;
+};
+typedef struct cjo_attribute_values_s cjo_attribute_values_t;
+/* }}} */
+
+struct cjo_membuffer_s {
+  char *buffer;
+  size_t size;
+  size_t used;
+};
+
+typedef struct cjo_membuffer_s cjo_membuffer_t;
+
+struct cjo_s /* {{{ */
+{
+  char *instance;
+  char *host;
+
+  char *sock;
+
+  char *url;
+  char *user;
+  char *pass;
+  char *credentials;
+
+  const cjo_bean_t *match_this_bean;
+  cjo_attribute_values_t *curr_attribute; /* points to the pool below. */
+  cjo_attribute_values_t *attributepool;
+  int attribute_pool_used;
+
+  const char *json_key;
+  yajl_len_t json_key_len;
+
+  cjo_expect_token expect;
+
+  _Bool verify_peer;
+  _Bool verify_host;
+  char *cacert;
+  struct curl_slist *headers;
+  char *post_body;
+  size_t post_body_len;
+  _Bool expect_escapes;
+  cdtime_t interval;
+  int timeout;
+
+  CURL *curl;
+  char curl_errbuf[CURL_ERROR_SIZE];
+  const unsigned char *start, *end;
+  cjo_membuffer_t replybuffer;
+  cjo_membuffer_t itembuffer;
+
+  size_t poolsize;
+  char *pool;
+  const char *poolmax;
+  char *pool_ptr;
+
+  yajl_handle yajl;
+  llist_t *bean_configs;
+  int max_attribute_count;
+  int max_value_len;
+};
+typedef struct cjo_s cjo_t; /* }}} */
+
+static void cjo_init_buffer(cjo_membuffer_t *buf, size_t initial_size) {
+  buf->buffer = malloc(initial_size);
+  buf->used = 0;
+  buf->size = initial_size;
+}
+
+static void cjo_append_buffer(cjo_membuffer_t *buf, const char *append,
+                              size_t appendlen) {
+  if (buf->used + appendlen + 1 > buf->size) {
+    char *newbuf;
```

Comment:
Please declare variables when initializing them, where possible. E.g.:

```c
size_t newsize = buf->size * 2;

char *newbuf = calloc(1, newsize);
```

Here and below.